Test/git-issues/git-issue-3804b.dfy

```diff
@@ -0,0 +1,42 @@
+// RUN: %exits-with 4 %baredafny verify %args "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+function FunctionAndMethodVerify(i: int): (r: int)
+  requires p: i >= 0
+  ensures r >= 0
+{
+  reveal p;
+  i
+} by method {
+  reveal p;
+  r := i;
```

Comment:
Note, this `by method` clause should also verify in the absence of L11. The reason is that the `by method` is verified to return the exact same value as the function body, and the function's `ensures` clauses are checked only of the function body.

Reply:
Hmm. Evidently, precondition labels are ignored during the well-formedness checking of specifications. Therefore, the use of `F` in the postcondition of method `M` above passes the well-formedness check, and so do the well-formedness checks of the specification in

``` dafny
method M'(i: int) returns (r: int)
  requires p: i >= 0
  requires 2 < 10 / if i >= 0 then 1 else 0
  ensures r == F(i)
```

It would seem consistent to say that the use of the function body in a `by method` still needs to prove the precondition of that function. But since that's the very same precondition that we're copying into the method that stands for the `by method` part, it seems that any reveals of the function's precondition ought to come for free. 

Would it be possible in our implementation (of the well-formedness check of `by method`) to `assume CanCall$F(i)` before checking the `r == F(i)` postcondition?

Reply:
Looks like it should work. I'll implement it.

Reply:
There is something I don't understand.

I changed the code a bit so that the by method declaration contains a conditional return
```dafny
function FunctionVerifiesButNotMethod(i: int): (r: int)
  requires p: i >= 0
  ensures r >= 0
{
  reveal p;
  i
} by method { // Error: Could not verify + related
  if(i == 5) {
    return  ; // Ensure we check every branch.
  }
  r := i;
}
```

Now, things start to get weird. In the generated boogie code, we have (I removed irrelevant pieces of code), I get an error on the last return path, but not the inner one.
```
procedure {:verboseName "FunctionVerifiesButNotMethod (correctness)"} Impl$$_module.__default.FunctionVerifiesButNotMethod(i#0: int) returns (defass#r#0: bool, r#0: int, $_reverifyPost: bool);
  // user-defined postconditions
  free ensures _module.__default.FunctionVerifiesButNotMethod#canCall(i#0);
  ensures r#0 == _module.__default.FunctionVerifiesButNotMethod(i#0);

implementation {:verboseName "FunctionVerifiesButNotMethod (correctness)"} Impl$$_module.__default.FunctionVerifiesButNotMethod(i#0: int) returns (defass#r#0: bool, r#0: int, $_reverifyPost: bool)
{
    $_reverifyPost := false;
    if (i#0 == LitInt(5))
    {
        r#0 := i#0;
        defass#r#0 := true;
        assert defass#r#0;
        return;
    }
    r#0 := i#0;
    defass#r#0 := true;
    assert defass#r#0;
} // Error here, cannot prove the precondition r#0 == _module.__default.FunctionVerifiesButNotMethod(i#0);
```
However, if I inline the free ensures as an assume and the ensures as an assert, Boogie is happy.
So it looks like there is actually a problem on the Boogie side.